config/jobs/kubernetes/sig-testing/apidiff.yaml

```diff
@@ -56,5 +56,82 @@ presubmits:
             cpu: 2
             memory: 12Gi
 
-# A periodic job which shows API diffs for staging repos since the last release
-# might be useful. Not done yet.
+  - name: pull-kubernetes-apidiff-client-go
+    cluster: eks-prow-build-cluster
+    # A job which automatically runs for changes in client-go or the generated code
+    # to have visibility on the changes that will impact the external projects
+    # that are using the Kubernetes golang clients
+    run_if_changed: '(^staging\/src\/k8s.io\/client-go)|(^staging\/src\/k8s.io\/code-generator\/examples)'
+    optional: true
+    decorate: true
+    annotations:
+      # The apidiff.sh script uses the latest revision of apidiff.
+      # There is no guarantee that this will continue to work for
+      # older branches, so let's not even create per-release
+      # copies of this job.
+      fork-per-release: "false"
+      testgrid-dashboards: sig-testing-misc
+      testgrid-create-test-group: 'true'
+    path_alias: k8s.io/kubernetes
+    spec:
+      containers:
+      - image: gcr.io/k8s-staging-test-infra/kubekins-e2e:v20241128-8df65c072f-master
+        command:
+        - runner.sh
+        args:
+        - /bin/sh
+        - -c
+        # PULL_BASE_SHA is the revision on the target branch that the
+        # PR gets merged into. What we want instead is the revision at
+        # which the PR branch diverged from the target branch.
+        # We get that from "git merge-base".
+        - "./hack/apidiff.sh -r $(git merge-base ${PULL_BASE_SHA} ${PULL_PULL_SHA}) -t ${PULL_PULL_SHA} ./staging/src/k8s.io/code-generator/examples ./staging/src/k8s.io/client-go"
```

Comment:
I recommend moving this fully out into a script somewhere, it will need iterating on and inline in prowjob config is a poor place to iterate on scripts.

I don't consider this a hard blocker, but seriously consider doing this sooner than later.
If it were planned to be a blocking job, I'd be more concerned.

The moment this apidiff script wants more computed inputs, we have a mess, but if some wrapper script takes in the prow envs, it's not a problem.

Reply:
I don't think adding an indirection in this case is going to help, if we end needing more complexity for this job then we are doing something wrong

Reply:
usually the prowjob should make a trivial invocation like `make test`, and encapsulate the rest in the repo under test, where you can iterate on it with a tested PR

it's not about _more_ complexity, it's about any change to how this needs to be invoked. right now you're passing quite a lot as arguments coming from env or even computed.

Reply:
(which also makes local replication more difficult)

Reply:
🤔  there are only 2 arguments that depend on prow variables

Reply:
We could add a `hack/prow-apidiff.sh` which then would contain roughly:
```
apidiff.sh -r $(git merge-base ${PULL_BASE_SHA} ${PULL_PULL_SHA}) -t ${PULL_PULL_SHA} "$@"
```

Is that worth it? If we need it in multiple jobs, then perhaps. :shrug: 

Normal developers won't need to call it. To check their local branch, `hack/apidiff.sh <directories>` without `-r` and `-t` will already figure out the revisions automatically. The situation in Prow is more complicated because there's no branch information and additional merge commits.

Reply:
It is easier for a contributor to understand "ah to mimic this locally I should just run `apidiff.sh`" if that's what CI invokes.

Reply:
(aside: PULL_PULL_SHA will not be set on batch PRs, TODO to rework this out if we ever want a blocking job later)

Reply:
`apidiff.sh` was meant to be agnostic of the CI system in which it gets run. But I don't feel strongly about this.

> (aside: PULL_PULL_SHA will not be set on batch PR runs, TODO to rework this out if we ever want a blocking job later)

It's complexity like this that I wanted to keep out of the script... In the job it's clearer which variables are set. :shrug:

Reply:
https://github.com/kubernetes/kubernetes/pull/129188

Reply:
> apidiff.sh was meant to be agnostic of the CI system in which it gets run. But I don't feel strongly about this.

... eh... prow is the CI system Kubernetes uses, we normally support it in our tools / scripts.

> It's complexity like this that I wanted to keep out of the script... In the job it's clearer which variables are set. 🤷

That complexity has to go somewhere, and it's harder to iterate on correctly if it's inline in the job.

Reply:
(also, we're not _requiring_ prow-like-environment-variables, just supporting it as a fallback if the args are not otherwise set, see the PR)